deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/HostTerminator.java

```diff
@@ -29,31 +33,38 @@
 
 public class HostTerminator implements Runnable {
     private static final Logger LOG = LoggerFactory.getLogger(HostTerminator.class);
-    private static final long timeToRetain = 60 * 60 * 1000; // 1 hour
+    private final AgentDAO agentDAO;
     private final HostDAO hostDAO;
     private final HostInfoDAO hostInfoDAO;
     private final UtilDAO utilDAO;
+    private final ClusterHandler clusterHandler;
 
     public HostTerminator(ServiceContext serviceContext) {
+        agentDAO = serviceContext.getAgentDAO();
         hostDAO = serviceContext.getHostDAO();
         hostInfoDAO = serviceContext.getHostInfoDAO();
         utilDAO = serviceContext.getUtilDAO();
+        clusterHandler = new ClusterHandler(serviceContext);
     }
 
-    void terminateHost(HostBean host) throws Exception {
-        if ((System.currentTimeMillis() - host.getLast_update()) > timeToRetain) {
-            String hostId = host.getHost_id();
-            LOG.info(String.format("This host %s has been retained for more than 1 hours since it's last updated. Terminate it", hostId));
-            hostInfoDAO.terminateHost(hostId);
+    private void terminateHost(HostBean host) throws Exception {
+        String hostId = host.getHost_id();
+        List<AgentBean> agentBeans = agentDAO.getByHostId(hostId);
+        boolean stopSucceeded = true;
+        for (AgentBean agentBean : agentBeans) {
+            if (agentBean.getDeploy_stage() != DeployStage.STOPPED) {
+                stopSucceeded = false;
+            }
+        }
 
-            HostBean hostBean = new HostBean();
-            hostBean.setState(HostState.TERMINATING);
-            hostBean.setLast_update(System.currentTimeMillis());
-            hostDAO.updateHostById(hostId, hostBean);
+        if (stopSucceeded) {
+            LOG.info(String.format("Host %s is stopped. Terminate it.", hostId));
+            String clusterName = host.getGroup_name();
+            clusterHandler.terminateHosts(clusterName, Arrays.asList(hostId));
```

Comment:
nitpick: you can use Collections.singlelist if it's immutable:
http://stackoverflow.com/questions/26027396/arrays-aslist-vs-collections-singletonlist